packages/labs/signals/src/lib/signal-watcher.ts

```diff
@@ -0,0 +1,176 @@
+        this.__watcher!.unwatch(this.__performUpdateSignal!);
+        this.__performUpdateSignal = undefined;
+        this.__watcher = undefined;
+      }
+    }
+
+    /**
+     * Used to force an uncached read of the __updateSignal when we need to read
+     * the current value during an update.
+     *
+     * If https://github.com/tc39/proposal-signals/issues/151 is resolved, we
+     * won't need this.
+     */
+    private __forceUpdateSignal = new Signal.State(0);
+
+    private __forcingUpdate = false;
+
+    /**
+     * A computed signal that wraps performUpdate() so that all signals that are
+     * accessed during the update lifecycle are tracked.
+     *
+     * __forceUpdateSignal is used to force an uncached read of this signal
+     * because updates may easily depend on non-signal values, so we must always
+     * re-run it.
+     */
+    private __performUpdateSignal?: Signal.Computed<void>;
+
+    /**
+     * Whether or not the next update should perform a full render, or if only
+     * pending watches should be committed.
+     *
+     * If requestUpdate() was called only because of watch() directive updates,
+     * then we can just commit those directives without a full render. If
+     * requestUpdate() was called for any other reason, we need to perform a
+     * full render, and don't need to separately commit the watch() directives.
+     *
+     * This is set to `true` initially, and whenever requestUpdate() is called
+     * outside of a watch() directive update. It is set to `false` when
+     * update() is called, so that a requestUpdate() is required to do another
+     * full render.
+     */
+    private __doFullRender = true;
+
+    /**
+     * Set of watch directives that have been updated since the last update.
+     * These will be committed in update() to ensure that the latest value is
+     * rendered and that all updates are batched.
+     */
+    private __pendingWatches = new Set<WatchDirective<unknown>>();
+
+    protected override performUpdate() {
+      // Always enable watching before an update, even if disconnected, so that
+      // we can track signals that are accessed during the update.
+      this.__watch();
+      // Force an uncached read of __performUpdateSignal
+      this.__forcingUpdate = true;
+      this.__forceUpdateSignal?.set(this.__forceUpdateSignal.get() + 1);
+      this.__forcingUpdate = false;
+      // Always read from the signal to ensure that it's tracked
+      this.__performUpdateSignal!.get();
+    }
+
+    protected override update(
+      changedProperties: PropertyValueMap<this> | Map<PropertyKey, unknown>
+    ): void {
+      if (this.__doFullRender) {
+        // Force future updates to not perform full renders by default.
+        this.__doFullRender = false;
+        super.update(changedProperties);
+      } else {
+        // For a partial render, just commit the pending watches.
+        this.__pendingWatches.forEach((d) => d.commmit());
```

Comment:
Note, this performs the commits in directive update order. We probably want to do this is tree order.

Reply:
That does seem a lot better because it would ensure that a signal referenced by a child doesn't do unnecessary work that a parent would make superfluous (like DOM removal).

I'm not sure how to do it efficiently, and let's definitely leave a comment if we don't resolve it. Possibly we could leverage the fact that any full render will be in order and push onto an array on the host in that case. Then this ordered array would always be iterated here and only pending watches would be executed.

Reply:
```suggestion
        this.__pendingWatches.forEach((d) => d.commit());
```